File: nf-backend/src/main/java/com/neuroforge/backend/project/service/TaskService.java
package com.neuroforge.backend.project.service;

import com.neuroforge.backend.dto.ApiResponse;
import com.neuroforge.backend.project.dto.CreateTaskRequest;
import com.neuroforge.backend.project.dto.TaskBoardDto;
import com.neuroforge.backend.project.dto.TaskDto;
import com.neuroforge.backend.project.dto.TaskStatusHistoryResponse;
import com.neuroforge.backend.project.dto.UpdateTaskRequest;
import com.neuroforge.backend.project.dto.UpdateTaskStatusRequest;

import java.util.List;

public interface TaskService {

    ApiResponse<TaskDto> createTask(CreateTaskRequest request);

    ApiResponse<List<TaskDto>> getAllTasks();

    ApiResponse<List<TaskDto>> getTasksByProject(Long projectId);

    ApiResponse<List<TaskDto>> getTasksBySprint(Long sprintId);

    // Module 3: task board view grouped by status
    ApiResponse<TaskBoardDto> getTaskBoard(Long projectId);

    ApiResponse<TaskDto> getTaskById(Long id);

    ApiResponse<TaskDto> updateTask(Long id, UpdateTaskRequest request);

    ApiResponse<Void> deleteTask(Long id);

    // Module 5: Task status update with history tracking
    ApiResponse<TaskDto> updateTaskStatus(Long taskId, UpdateTaskStatusRequest request);

    // Module 5: Get task status history
    ApiResponse<List<TaskStatusHistoryResponse>> getTaskStatusHistory(Long taskId);
}

File: neuroforge-frontend/src/services/taskService.ts
import api from './api';

// ─── Types ────────────────────────────────────────────────────────────────────

export enum TaskStatus {
  TODO = 'TODO',
  IN_PROGRESS = 'IN_PROGRESS',
  CODE_REVIEW = 'CODE_REVIEW',
  TESTING = 'TESTING',
  DONE = 'DONE'
}

export enum TaskPriority {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL'
}

export interface Task {
  id: string;
  title: string;
  description?: string;
  status: TaskStatus;
  priority: TaskPriority;
  storyPoints?: number;
  labels?: string;
  // Module 4: Specification Traceability
  specificationId?: string;
  specificationVersionId?: string;
  specificationTitle?: string;
  specificationVersionNumber?: number;
  sprintId?: string;
  assigneeId?: string;
  createdAt?: string;
  createdBy?: string;
  updatedAt?: string;
  updatedBy?: string;
}

export interface CreateTaskRequest {
  title: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  storyPoints?: number;
  labels?: string;
  // Module 4: Specification Traceability (optional)
  specificationId?: string;
  specificationVersionId?: string;
  sprintId?: string;
  assigneeId?: string;
}

export interface UpdateTaskStatusRequest {
  status: TaskStatus;
}

export interface TaskStatusHistory {
  id: string;
  taskId: string;
  oldStatus?: TaskStatus;
  newStatus: TaskStatus;
  changedAt: string;
  changedBy?: string;
}

// ─── Service ──────────────────────────────────────────────────────────────────

export const taskService = {
  // ── Task CRUD ────────────────────────────────────────────────────────────
  createTask: (data: CreateTaskRequest) => api.post<Task>('/tasks', data),
  getAllTasks: () => api.get<Task[]>('/tasks'),
  getBacklogTasks: () => api.get<Task[]>('/tasks/backlog'),
  searchTasks: (keyword?: string) => api.get<Task[]>(`/tasks/search${keyword ? `?keyword=${keyword}` : ''}`),
  getTasksByStatus: (status: TaskStatus) => api.get<Task[]>(`/tasks/status/${status}`),
  getTasksByPriority: (priority: TaskPriority) => api.get<Task[]>(`/tasks/priority/${priority}`),
  getTasksBySprint: (sprintId: string) => api.get<Task[]>(`/tasks/sprint/${sprintId}`),
  getTasksByAssignee: (userId: string) => api.get<Task[]>(`/tasks/assignee/${userId}`),
  getTaskById: (id: string) => api.get<Task>(`/tasks/${id}`),
  updateTask: (id: string, data: CreateTaskRequest) => api.put<Task>(`/tasks/${id}`, data),
  deleteTask: (id: string) => api.delete(`/tasks/${id}`),

  // ── Task Assignment ────────────────────────────────────────────────────────
  assignSprint: (taskId: string, sprintId: string) => api.patch<Task>(`/tasks/${taskId}/assign-sprint/${sprintId}`),
  removeSprint: (taskId: string) => api.patch<Task>(`/tasks/${taskId}/remove-sprint`),
  assignUser: (taskId: string, userId: string) => api.patch<Task>(`/tasks/${taskId}/assign-user/${userId}`),
  removeUser: (taskId: string) => api.patch<Task>(`/tasks/${taskId}/remove-user`),

  // ── Task Status ───────────────────────────────────────────────────────────
  updateTaskStatus: (taskId: string, data: UpdateTaskStatusRequest) => api.patch<Task>(`/tasks/${taskId}/status`, data),

  // ── Task History ───────────────────────────────────────────────────────────
  getTaskStatusHistory: (taskId: string) => api.get<TaskStatusHistory[]>(`/tasks/${taskId}/history`),
};
